llm-examples/mpt/mpt-7b/scripts/fine_tune_deepspeed.py

```diff
@@ -0,0 +1,261 @@
+
+    return train_tokenized_dataset, eval_tokenized_dataset
+      
+
+
+def load_model(
+    pretrained_model_name_or_path: str = MODEL_PATH,
+) -> AutoModelForCausalLM:
+    logger.info(f"Loading model for {pretrained_model_name_or_path}")
+    config = AutoConfig.from_pretrained(pretrained_model_name_or_path,
+                                        trust_remote_code=True
+                                        )
+    config.attn_config['attn_impl'] = 'torch'
+    # config.init_device = 'cuda'
+    model_hidden_size = config.d_model
+
+    model = transformers.AutoModelForCausalLM.from_pretrained(
+      pretrained_model_name_or_path,
+      config=config,
+      torch_dtype=torch.bfloat16,
+      trust_remote_code=True
+    )
+    return model, model_hidden_size
+
+def get_tokenizer(
+    pretrained_tokenizer_name_or_path: str = TOKENIZER_PATH,
+) -> PreTrainedTokenizer:
+    tokenizer = AutoTokenizer.from_pretrained(pretrained_tokenizer_name_or_path)
+    tokenizer.pad_token = tokenizer.eos_token
+    return tokenizer
+
+with open(CONFIG_PATH) as config_json:
+  ds_config_dict = json.load(config_json)
+
+def train(
+  *,
+  input_model: str,
+  local_output_dir: str,
+  dbfs_output_dir: str,
+  epochs: int,
+  per_device_train_batch_size: int,
+  per_device_eval_batch_size: int,
+  lr: float,
+  seed: int,
+  gradient_checkpointing: bool,
+  local_rank: str,
+  bf16: bool,
+  logging_steps: int,
+  save_steps: int,
+  eval_steps: int,
+  save_total_limit: int,
+  warmup_steps: int,
+  training_dataset: str = DEFAULT_TRAINING_DATASET,
+):
+  set_seed(seed)
+  os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "max_split_size_mb:128"
+  # Enable tf32 for better performance
+  torch.backends.cuda.matmul.allow_tf32 = True
+
+  model, model_hidden_size = load_model(pretrained_model_name_or_path=input_model)
+  tokenizer = get_tokenizer()
+  train_dataset, val_dataset = load_training_dataset(tokenizer, seed=seed)
+  ds_config_dict["hidden_size"] = model_hidden_size
+  ds_config_dict["zero_optimization"]["reduce_bucket_size"] = model_hidden_size*model_hidden_size
+  ds_config_dict["zero_optimization"]["stage3_prefetch_bucket_size"] = 0.9 * model_hidden_size * model_hidden_size
+  ds_config_dict["zero_optimization"]["stage3_param_persistence_threshold"] = 10 * model_hidden_size
+
+  # enable fp16 if not bf16
+  fp16 = not bf16
+
+  training_args = TrainingArguments(
+    output_dir=local_output_dir,
+    per_device_train_batch_size=per_device_train_batch_size,
+    per_device_eval_batch_size=per_device_eval_batch_size,
+    learning_rate=lr,
+    num_train_epochs=epochs,
+    weight_decay=1,
+    do_eval=True,
+    evaluation_strategy="steps",
+    eval_steps=eval_steps,
+    fp16=fp16,
+    bf16=bf16,
+    deepspeed=ds_config_dict,
+    logging_strategy="steps",
+    logging_steps=logging_steps,
+    save_strategy="steps",
+    save_steps=save_steps,
+    save_total_limit=save_total_limit,
+    max_steps=20,
```

Comment:
Should we also expose `max_steps` as an argument or leave a comment that it is encouraged to increase this for custom dataset, since 20 is pretty small for fine-tuning?

Reply:
Thanks. I will delete this before merge so that it will train for whole epochs.